### ISSUE TITLE
If I cancel the download, it automatically restarts.

### ISSUE THREAD

**Issue by Nahid-mk** [781flyingdutchman/background_downloader#404]
**Describe the bug**
When I cancel the download, it doesn't stay canceled; instead, it starts up again automatically on its own. This happens without any action on my part, as if the system is programmed to resume the download immediately whenever I try to stop it. So, even if I cancel intentionally, it restarts itself without any prompt or confirmation. The problem is it's not happening  every time.

**Expected behavior**
It should be canceled when ever i call the cancelDownload() functioin.

**Logs**
V/BackgroundDownloader(21429): Canceling taskIds [903986795]
D/BackgroundDownloader(21429): Could not find task with taskId 903986795 to cancel
I/WM-WorkerWrapper(21429): Work [ id=a87d0c2d-9459-4152-8c53-a7cd6503b71f, tags={ com.bbflight.background_downloader.DownloadTaskWorker, BackgroundDownloader, taskId=903986795, group=default } ] was cancelled
I/WM-WorkerWrapper(21429): java.util.concurrent.CancellationException: Task was cancelled.
I/WM-WorkerWrapper(21429): 	at androidx.work.impl.utils.futures.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1183)
I/WM-WorkerWrapper(21429): 	at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:513)
I/WM-WorkerWrapper(21429): 	at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:474)
I/WM-WorkerWrapper(21429): 	at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:316)
I/WM-WorkerWrapper(21429): 	at androidx.work.impl.utils.SerialExecutorImpl$Task.run(SerialExecutorImpl.java:96)
I/WM-WorkerWrapper(21429): 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
I/WM-WorkerWrapper(21429): 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
I/WM-WorkerWrapper(21429): 	at java.lang.Thread.run(Thread.java:1012)
I/TaskWorker(21429): Exception for taskId 903986795: kotlinx.coroutines.JobCancellationException: Job was cancelled; job=JobImpl{Cancelling}@898a82c
D/BackgroundDownloader(21429): Could not find task with taskId 903986795 to cancel
D/BackgroundDownloader(21429): Could not find task with taskId 903986795 to cancel
D/BackgroundDownloader(21429): Could not find task with taskId 903986795 to cancel
D/BackgroundDownloader(21429): Could not find task with taskId 903986795 to cancel

**Code**
 Future<void> cancelDownload({required String taskId}) async {
    await FileDownloader().cancelTaskWithId(taskId);
  }
  
  
   void enableListener() {
    FileDownloader().updates.listen((update) {
      String? id = controller.taskMapping[update.task.taskId];
      if (id != null) {
        switch (update) {
          case TaskStatusUpdate _:
            switch (update.status) {
              case TaskStatus.complete:
                controller.downloadedVideo
                    .add("$id.${Platform.isIOS ? "mp4" : "oob"}");
                controller.onGoingDownload.remove(id);
                controller.taskMapping.remove(update.task.taskId);
                break;

              case TaskStatus.canceled:
                controller.onGoingDownload.remove(id);
                controller.taskMapping.remove(update.task.taskId);
                break;

              case TaskStatus.enqueued:
                controller.onGoingDownload[id] = 0;
                showMassage("Download started in background.", isSuccess: true);
                break;

              case TaskStatus.running:
                controller.onGoingDownload[id] = 0;
                showMassage("Download is starting", isSuccess: true);
                break;
              default:
            }
            break;

          case TaskProgressUpdate _:
            controller.onGoingDownload[id] = update.progress;
            break;
        }
      }
    });
  }

**Comment by 781flyingdutchman** [comment]
Can you elaborate on when this happens? From the logs it looks like the downloader cannot find the task with the given id, which is unusual, so I want to understand if this happens for example when the task is cancelled from a notification, or after a long time

**Comment by Nahid-mk** [comment]
I implemented a download cancellation feature in my Flutter app. When a download starts, I store the task ID so I can cancel it if needed. I added a cancel button that calls the following function, passing in the stored task ID to cancel the download:

Future<void> cancelDownload({required String taskId}) async {
  await FileDownloader().cancelTaskWithId(taskId);
}
When I press the cancel button during a download, this function is triggered to cancel the task associated with the specified task ID."

**Comment by 781flyingdutchman** [comment]
Thanks - you said it's not happening every time, so can you explain when it dies and when it doesn't happen? Also, can you confirm Android version, phone model and plug-in version?

**Comment by 781flyingdutchman** [comment]
OK, can you add a log specific to that situation, and mark in it where you turn off WiFi, where you turn on WiFi and when you press the cancel button? Also please answer my other questions - I need this info to help solve this issue.

**Comment by Nahid-mk** [comment]
I start downloading a video. After 10-12% of the download completes, I turn off the Wi-Fi. Then, I wait for about 40 seconds and turn the Wi-Fi back on. The download restarts from the beginning, and after reaching 5-8%, I press the cancel button. This is the scenario where the issue occurs.

**Comment by 781flyingdutchman** [comment]
Thanks, but I really need to see the logs, and in those logs just indicate each of these events (like when you turn off the wifi, when you turn it back on, etc).

**Comment by 781flyingdutchman** [comment]
@Soap-141 does this happen only in the scenario that @Nahid-mk describes, i.e. after interrupting WiFi?  Can you and @Nahid-mk both share full logs of when this happens? Thanks.

**Comment by Soap-141** [comment]
@781flyingdutchman When simply cancelling a task using a group or ID.

**Comment by Soap-141** [comment]
I ended up handling the retry logic myself.

**Comment by 781flyingdutchman** [comment]
For my own refrerence:
```
>>> Disconnect WiFi
Work [ id=a30d3ef3-9e67-47ff-934a-4b3cbade13d0, tags={ com.bbflight.background_downloader.DownloadTaskWorker, BackgroundDownloader, taskId=327404472, group=default } ] was cancelled
java.util.concurrent.CancellationException: Task was cancelled.
	at androidx.work.impl.utils.futures.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1183)
	at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:513)
	at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:474)
	at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:316)
	at androidx.work.impl.utils.SerialExecutorImpl$Task.run(SerialExecutorImpl.java:96)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
	at java.lang.Thread.run(Thread.java:1012)
Exception for taskId 327404472: kotlinx.coroutines.JobCancellationException: Job was cancelled; job=JobImpl{Cancelling}@c25d1a4
Canceled task with id 327404472: Job was cancelled

>>> Reconnect WiFI
Starting task with taskId 327404472

>>> Cancel manually
Canceling taskIds [327404472]
Could not find task with taskId 327404472 to cancel
Work [ id=a30d3ef3-9e67-47ff-934a-4b3cbade13d0, tags={ com.bbflight.background_downloader.DownloadTaskWorker, BackgroundDownloader, taskId=327404472, group=default } ] was cancelled
java.util.concurrent.CancellationException: Task was cancelled.
	at androidx.work.impl.utils.futures.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1183)
	at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:513)
	at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:474)
	at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:316)
	at androidx.work.impl.utils.SerialExecutorImpl$Task.run(SerialExecutorImpl.java:96)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
	at java.lang.Thread.run(Thread.java:1012)

>>> 2 seconds later
Enqueuing task with id 327404472
Resuming task with taskId 327404472

>>> Cancelling manually again
Canceling taskIds [327404472]
Could not find task with taskId 327404472 to cancel
Work [ id=963564b7-e74a-4de3-9077-f95a3e409b21, tags={ com.bbflight.background_downloader.DownloadTaskWorker, BackgroundDownloader, taskId=327404472, group=default } ] was cancelled
java.util.concurrent.CancellationException: Task was cancelled.
	at androidx.work.impl.utils.futures.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1183)
	at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:513)
	at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:474)
	at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:316)
	at androidx.work.impl.utils.SerialExecutorImpl$Task.run(SerialExecutorImpl.java:96)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
	at java.lang.Thread.run(Thread.java:1012)
Exception for taskId 327404472: kotlinx.coroutines.JobCancellationException: Job was cancelled; job=JobImpl{Cancelling}@6ab872b
Canceled task with id 327404472: Job was cancelled
```

**Comment by 781flyingdutchman** [comment]
Logs for a normal cancellation:
```
Canceling taskIds [4143447675]
Work [ id=fa8a8657-a767-48eb-976e-5a4a2b9c5463, tags={ com.bbflight.background_downloader.DownloadTaskWorker, BackgroundDownloader, taskId=4143447675, group=default } ] was cancelled
java.util.concurrent.CancellationException: Task was cancelled.
	at androidx.work.impl.utils.futures.AbstractFuture.cancellationExceptionWithCause(AbstractFuture.java:1183)
	at androidx.work.impl.utils.futures.AbstractFuture.getDoneValue(AbstractFuture.java:513)
	at androidx.work.impl.utils.futures.AbstractFuture.get(AbstractFuture.java:474)
	at androidx.work.impl.WorkerWrapper$2.run(WorkerWrapper.java:316)
	at androidx.work.impl.utils.SerialExecutorImpl$Task.run(SerialExecutorImpl.java:96)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:644)
	at java.lang.Thread.run(Thread.java:1012)
Exception for taskId 4143447675: kotlinx.coroutines.JobCancellationException: Job was cancelled; job=JobImpl{Cancelling}@833d228
Canceled task with id 4143447675: Job was cancelled
```

**Comment by 781flyingdutchman** [comment]
Fixed in V8.8.0
